src/Cooked/MockChain/GenerateTx/Body.hs

```diff
@@ -97,17 +105,50 @@ txSkelToBodyContent skel@TxSkel {..} | txSkelReferenceInputs <- txSkelReferenceT
-    lift $
-      bimap
-        (TxBodyError "generateTx: ")
-        (`Cardano.Tx` [])
-        (Cardano.createAndValidateTransactionBody Cardano.ShelleyBasedEraConway txBodyContent)
-  foldM
-    ( \tx wal ->
-        case Ledger.addCardanoTxWitness (Ledger.toWitness $ Ledger.PaymentPrivateKey $ walletSK wal) (Ledger.CardanoTx tx Cardano.ShelleyBasedEraConway) of
-          Ledger.CardanoTx tx' Cardano.ShelleyBasedEraConway -> return tx'
-          _ -> throwOnString "txSkelToCardanoTx: Wrong output era"
-    )
-    cardanoTxUnsigned
-    (txSkelSigners txSkel)
+
+  -- We create the associated Shelley TxBody
+  txBody@(Cardano.ShelleyTxBody a body c dats e f) <-
+    lift $ mapLeft (TxBodyError "generateTx :") $ Cardano.createAndValidateTransactionBody Cardano.ShelleyBasedEraConway txBodyContent
+
+  -- There is a chance that the body is in need of additional data. This happens
+  -- when the set of reference inputs contains hashed datums that will need to
+  -- be resolved during phase 2 validation. All that follows until the
+  -- definition of "txBody'" aims at doing just that. In the process, we have to
+  -- reconstruct the body with the new data and the associated hash. Hopefully,
+  -- in the future, cardano-api provides a way to add those data in the body
+  -- directly without requiring this methods, which somewhat feels like a hack.
+  mData <- asks managedData
+  mTxOut <- asks managedTxOuts
+  refIns <- forM (txSkelReferenceTxOutRefs txSkel) $ \oRef ->
+    throwOnLookup ("txSkelToCardanoTx: Unable to resolve TxOutRef " <> show oRef) oRef mTxOut
+  let datumHashes = [hash | (Api.TxOut _ _ (Api.OutputDatumHash hash) _) <- refIns]
+  additionalData <- forM datumHashes $ \dHash ->
+    throwOnLookup ("txSkelToCardanoTx: Unable to resolve datum hash " <> show dHash) dHash mData
+  let additionalDataMap = Map.fromList [(Cardano.hashData dat, dat) | Api.Datum (Cardano.Data . Api.toData -> dat) <- additionalData]
+  toLangDepViewParam <- asks (Conway.getLanguageView . Cardano.unLedgerProtocolParameters . Emulator.ledgerProtocolParameters . params)
+  let txDats' = Alonzo.TxDats additionalDataMap
+      (era, datums, redeemers) = case dats of
+        Cardano.TxBodyNoScriptData -> (Cardano.AlonzoEraOnwardsConway, txDats', Alonzo.Redeemers Map.empty)
+        Cardano.TxBodyScriptData era' txDats reds -> (era', txDats <> txDats', reds)
+      witnesses = Cardano.collectTxBodyScriptWitnesses Cardano.ShelleyBasedEraConway txBodyContent
+      languages = [toCardanoLanguage v | (_, Cardano.AnyScriptWitness (Cardano.PlutusScriptWitness _ v _ _ _ _)) <- witnesses]
+      scriptIntegrityHash =
+        Cardano.alonzoEraOnwardsConstraints era $
+          Alonzo.hashScriptIntegrity (Set.fromList $ toLangDepViewParam <$> languages) redeemers datums
+      body' = body Lens.& Alonzo.scriptIntegrityHashTxBodyL Lens..~ scriptIntegrityHash
+      txBody' = Cardano.ShelleyTxBody a body' c (Cardano.TxBodyScriptData era datums redeemers) e f
```

Comment:
I think this is a bit rough, it would be nice to add a few comments to describe the steps in natural language and avoid brain time parsing what happens when just taking a glance.

Reply:
I commented it line by line.